bundles/org.openhab.binding.salus/src/main/java/org/openhab/binding/salus/internal/rest/AbstractSalusApi.java

```diff
@@ -0,0 +1,114 @@
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.salus.internal.SalusApi;
+import org.openhab.binding.salus.internal.rest.exceptions.AuthSalusApiException;
+import org.openhab.binding.salus.internal.rest.exceptions.SalusApiException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Martin Grześlowski - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractSalusApi<AuthT> implements SalusApi {
+    private static final int MAX_RETRIES = 3;
+    protected static final long TOKEN_EXPIRE_TIME_ADJUSTMENT_SECONDS = 3;
+    protected final Logger logger;
+    protected final String username;
+    protected final byte[] password;
+    protected final String baseUrl;
+    protected final RestClient restClient;
+    protected final GsonMapper mapper;
+    @Nullable
+    protected ZonedDateTime authTokenExpireTime;
+    protected final Clock clock;
+    @Nullable
+    protected AuthT authentication;
+
+    protected AbstractSalusApi(String username, byte[] password, String baseUrl, RestClient restClient,
+            GsonMapper mapper, Clock clock) {
+        this.username = username;
+        this.password = password;
+        this.baseUrl = removeTrailingSlash(baseUrl);
+        this.restClient = restClient;
+        this.mapper = mapper;
+        this.clock = clock;
+        // thanks to this, logger will always inform for which rest client it's doing the job
+        // it's helpful when more than one SalusApi exists
+        logger = LoggerFactory.getLogger(this.getClass().getName() + "[" + username.replace(".", "_") + "]");
+    }
+
+    public AbstractSalusApi(String username, byte[] password, String baseUrl, RestClient restClient,
+            GsonMapper mapper) {
+        this(username, password, baseUrl, restClient, mapper, Clock.systemDefaultZone());
+    }
+
+    protected @Nullable String get(String url, RestClient.Header... headers) throws SalusApiException {
+        refreshAccessToken();
+        return restClient.get(url, headers);
+    }
+
+    protected @Nullable String post(String url, RestClient.Content content, RestClient.Header... headers)
+            throws SalusApiException {
+        refreshAccessToken();
+        return restClient.post(url, content, headers);
+    }
+
+    private static String removeTrailingSlash(String str) {
+        if (str.endsWith("/")) {
+            return str.substring(0, str.length() - 1);
+        }
+        return str;
+    }
+
+    private void forceRefreshAccessToken() throws SalusApiException {
+        logger.debug("Force refresh access token");
+        cleanAuth();
+        refreshAccessToken();
+    }
+
+    protected synchronized final void refreshAccessToken() throws AuthSalusApiException {
+        if (this.authentication == null || isExpiredToken()) {
+            cleanAuth();
+            try {
+                login();
+            } catch (Exception ex) {
```

Comment:
login() only throws  `SalusApiException`. Better to catch that and nothing else.

Reply:
Not quite true. Look at AWS implementation. It uses some cognito staff (that underneath uses HTTP) that I have no idea what can it throw.

I can downgrade to `RuntimeEx`

Reply:
Ok, thanks for the details. Let's leave it as is now.